crates/red_knot_python_semantic/src/semantic_index/use_def/bitset.rs

```diff
@@ -0,0 +1,402 @@
+        }
+    }
+
+    /// Insert a value into the [`BitSet`].
+    ///
+    /// Return true if the value was newly inserted, false if already present.
+    pub(super) fn insert(&mut self, value: u32) -> bool {
+        if value >= Self::BITS {
+            self.overflow();
+        }
+        match self {
+            Self::Inline(blocks) => {
+                let value_usize = value as usize;
+                let (block, index) = (value_usize / 64, value_usize % 64);
+                let missing = blocks[block] & (1_u64 << index) == 0;
+                blocks[block] |= 1_u64 << index;
+                missing
+            }
+            Self::Heap(set) => set.insert(value),
+        }
+    }
+
+    /// Intersect in-place with another [`BitSet`].
+    pub(super) fn intersect(&mut self, other: &BitSet<B>) {
+        match (self, other) {
+            (Self::Inline(myblocks), Self::Inline(other_blocks)) => {
+                for i in 0..B {
+                    myblocks[i] &= other_blocks[i];
+                }
+            }
+            (Self::Heap(myset), Self::Heap(other_set)) => {
+                *myset = myset.intersection(other_set).copied().collect();
+            }
+            (me, other) => {
+                for value in other.iter() {
+                    me.insert(value);
+                }
+            }
+        }
+    }
+
+    /// Return an iterator over the values (in ascending order) in this [`BitSet`].
+    pub(super) fn iter(&self) -> BitSetIterator<'_, B> {
+        match self {
+            Self::Inline(blocks) => BitSetIterator::Inline {
+                blocks,
+                cur_block_index: 0,
+                cur_block: blocks[0],
+            },
+            Self::Heap(set) => BitSetIterator::Heap(set.iter()),
+        }
+    }
+}
+
+/// Iterator over values in a [`BitSet`].
+#[derive(Debug)]
+pub(super) enum BitSetIterator<'a, const B: usize> {
+    Inline {
+        /// The blocks we are iterating over.
+        blocks: &'a [u64; B],
+
+        /// The index of the block we are currently iterating through.
+        cur_block_index: usize,
+
+        /// The block we are currently iterating through (and zeroing as we go.)
+        cur_block: u64,
+    },
+    Heap(btree_set::Iter<'a, u32>),
+}
+
+impl<const B: usize> Iterator for BitSetIterator<'_, B> {
+    type Item = u32;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Self::Inline {
+                blocks,
+                cur_block_index,
+                cur_block,
+            } => {
+                while *cur_block == 0 {
+                    if *cur_block_index == B - 1 {
+                        return None;
+                    }
+                    *cur_block_index += 1;
+                    *cur_block = blocks[*cur_block_index];
+                }
+                let value = cur_block.trailing_zeros();
+                // reset the lowest set bit
+                *cur_block &= cur_block.wrapping_sub(1);
```

Comment:
I might check how Rust compiles each of these; I thought the `wrapping_sub(1)` trick might be a bit more efficient. But it probably doesn't matter. I agree your version is easier to understand, which is why I added the comment :)

Reply:
Going to leave this; the `wrapping_sub(1)` way looks a bit more efficient; possibly most importantly it doesn't have a data dependence on storing the lowest set bit first, so it can allow a bit more parallel execution. I think with the comment it's sufficiently understandable.